File: python/binary_subarrays_with_sum.py
class Solution:
    def numSubarraysWithSum(self, nums: list[int], goal: int) -> int:
        """time complexity: O(n)"""
        return self.helper(nums, goal) - self.helper(nums, goal - 1)

    def helper(self, nums: list[int], x: int) -> int:
        """Count numbers of subarrays where curr <= x"""
        if x < 0:
            return 0
        l: int = 0
        curr: int = 0
        result: int = 0
        for r in range(len(nums)):
            curr += nums[r]
            while curr > x:
                curr -= nums[l]
                l += 1
            result += r - l + 1
        return result


def test_numSubarraysWithSum_case_1():
    """This is a test case"""
    # arrange
    nums: list[int] = [1, 0, 1, 0, 1]
    goal: int = 2
    expected: int = 4

    # act
    actual = Solution().numSubarraysWithSum(nums, goal)

    # assert
    assert expected == actual


def test_numSubarraysWithSum_case_2():
    """This is a test case"""
    # arrange
    nums: list[int] = [0, 0, 0, 0, 0]
    goal: int = 0
    expected: int = 15

    # act
    actual = Solution().numSubarraysWithSum(nums, goal)

    # assert
    assert expected == actual

File: rust/src/code/binary_subarrays_with_sum.rs
struct Solution;

impl Solution {
    pub fn num_subarrays_with_sum(nums: Vec<i32>, goal: i32) -> i32 {
        return Self::helper(&nums, goal) - Self::helper(&nums, goal - 1);
    }

    fn helper(nums: &Vec<i32>, x: i32) -> i32 {
        if x < 0 {
            return 0;
        }
        let mut l: usize = 0;
        let mut curr: i32 = 0;
        let mut result: i32 = 0;
        for r in 0..nums.len() {
            curr += nums[r];
            while curr > x {
                curr -= nums[l];
                l += 1;
            }
            if r >= l {
                result += (r - l + 1) as i32;
            }
        }
        result
    }
}

#[cfg(test)]
mod test {
    use super::Solution;

    #[test]
    fn test_num_subarrays_with_sum_case_1() {
        // arrange
        let nums: Vec<i32> = vec![1, 0, 1, 0, 1];
        let goal: i32 = 2;
        let expected: i32 = 4;

        // act
        let actual = Solution::num_subarrays_with_sum(nums, goal);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_num_subarrays_with_sum_case_2() {
        // arrange
        let nums: Vec<i32> = vec![0, 0, 0, 0, 0];
        let goal: i32 = 0;
        let expected: i32 = 15;

        // act
        let actual = Solution::num_subarrays_with_sum(nums, goal);

        // assert
        assert_eq!(expected, actual);
    }
}
